File: LeetCode/383RansomNote/RansomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # 2 Hashmaps -> Time:O(n+m), Space:O(n+m)
        mag_dict = {}
        for i in magazine:
            if i not in mag_dict:
                mag_dict[i] = 0
            mag_dict[i] += 1
        ransom_dict = {}
        for i in ransomNote:
            if i not in ransom_dict:
                ransom_dict[i] = 0
            ransom_dict[i] += 1
        for i in ransom_dict:
            if i not in mag_dict:
                return False
            elif ransom_dict[i] > mag_dict[i]:
                return False
        return True

File: LeetCode/383RansomNote/RansomNote.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        // 2 hashmaps -> Time:O(n+m), Space:O(n+m)
        std::map<char, int> magazine_map;
        std::map<char, int> ransom_map;

        for (int i = 0; i < magazine.size(); i++) {
            magazine_map[magazine[i]]++;
        }
        for (int i = 0; i < ransomNote.size(); i++) {
            ransom_map[ransomNote[i]]++;
        }

        for (auto i:ransom_map) {
            if (magazine_map.find(i.first) == magazine_map.end()) {
                return false;
            }
            else if (ransom_map[i.first] > magazine_map[i.first]) {
                return false;
            }
        }

        return true;
    }
};
